fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/StandardIndexMaintainer.java

```diff
@@ -392,22 +396,50 @@ public void updateUniquenessViolations(@Nonnull Tuple valueKey, @Nonnull Tuple p
         return scan(IndexScanType.BY_VALUE, TupleRange.ALL, continuation, scanProperties)
                 .filterAsync(
                         indexEntry -> state.store
                                 .hasIndexEntryRecord(indexEntry, IsolationLevel.SNAPSHOT)
                                 .thenApply(has -> !has),
-                        FDBRecordStore.DEFAULT_PIPELINE_SIZE)
-                .map(indexEntry ->
-                        new InvalidIndexEntry(indexEntry, InvalidIndexEntry.Reasons.ORPHAN));
+                        DEFAULT_PIPELINE_SIZE)
+                .map(InvalidIndexEntry::newOrphan);
+    }
+
+    /**
+     * Validate entries in the index. It scans the records and checks if the index entries associated with each record
+     * exist. Note that it may not work for synthetic indexes.
+     * @param continuation any continuation from a previous validation invocation
+     * @param scanProperties skip, limit and other properties of the validation (use default values if <code>null</code>)
+     * @return a cursor over records that have no associated index entries including the reason
+     */
+    @Nonnull
+    protected RecordCursor<InvalidIndexEntry> validateMissingEntries(@Nullable byte[] continuation,
+                                                                     @Nonnull ScanProperties scanProperties) {
+        final Collection<RecordType> recordTypes = state.store.getRecordMetaData().recordTypesForIndex(state.index);
+        return RecordCursor.flatMapPipelined(
+                cont -> state.store.scanRecords(TupleRange.ALL, cont, scanProperties)
+                        .filter(rec -> recordTypes.contains(rec.getRecordType())),
+                (record, cont) -> {
+                    List<IndexEntry> filteredIndexEntries = filteredIndexEntries(record);
+                    return RecordCursor.fromList(filteredIndexEntries == null ? Collections.emptyList() :
+                            filteredIndexEntries.stream()
+                                    .map(indexEntryWithoutPrimaryKey -> new IndexEntry(
+                                            indexEntryWithoutPrimaryKey.getIndex(),
+                                            indexEntryKey(indexEntryWithoutPrimaryKey.getKey(), record.getPrimaryKey()),
+                                            indexEntryWithoutPrimaryKey.getValue()
+                                    ))
+                                    .map(indexEntry -> Pair.of(indexEntry, record))
+                                    .collect(Collectors.toList()),
+                            cont);
+                },
+                continuation, DEFAULT_PIPELINE_SIZE)
+        .filterAsync(indexEntryRecordPair -> {
+            final byte[] keyBytes = state.indexSubspace.pack(indexEntryRecordPair.getLeft().getKey());
+            return state.transaction.get(keyBytes).thenApply(Objects::isNull);
+        }, DEFAULT_PIPELINE_SIZE)
```

Comment:
Given that this is reading records and looking for entries, maybe this shouldn't just use the default pipeline size and should instead have it configurable.